src/hotspot/share/opto/escape.cpp

```diff
@@ -4436,6 +4452,24 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
       _compile->get_alias_index(tinst->add_offset(oopDesc::mark_offset_in_bytes()));
       _compile->get_alias_index(tinst->add_offset(oopDesc::klass_offset_in_bytes()));
       if (alloc->is_Allocate() && (t->isa_instptr() || t->isa_aryptr())) {
+        // Update adr type captured by NarrowMem projections to new type
+        InitializeNode* init = alloc->as_Allocate()->initialization();
+        assert(init != nullptr, "can't find Initialization node for this Allocate node");
+        for (DUIterator_Fast imax, i = init->fast_outs(imax); i < imax; i++) {
+          ProjNode* proj = init->fast_out(i)->as_Proj();
+          if (proj->Opcode() == Op_NarrowMemProj) {
```

Comment:
```suggestion
          ProjNode* proj = init->fast_out(i)->as_NarrowMemProj();
          if (proj != nullptr) {
```
That way you don't need to do the hacky cast below `((NarrowMemProjNode*)proj)`.